Check types for privacy

petrochenkov opened the pull request:
This PR implements late post factum checking of type privacy, as opposed to early preventive "private-in-public" checking.
This will allow to turn private-in-public checks into a lint and make them more heuristic-based, and more aligned with what people may expect (e.g. reachability-based behavior).

Types are privacy-checked if they are written explicitly, and also if they are inferred as expression or pattern types.
This PR checks "semantic" types and does it unhygienically, this significantly restricts what macros 2.0 (as implemented in https://github.com/rust-lang/rust/pull/40847) can do (sorry @jseyfried) - they still can use private *names*, but can't use private *types*.
This is the most conservative solution, but hopefully it's temporary and can be relaxed in the future, probably using macro contexts of expression/pattern spans.

Traits are also checked in preparation for [trait aliases](https://github.com/rust-lang/rust/issues/41517), which will be able to leak private traits, and macros 2.0 which will be able to leak pretty much anything.

This is a [breaking-change], but the code that is not contrived and can be broken by this patch should be guarded by `private_in_public` lint. [Previous crater run](https://github.com/rust-lang/rust/issues/34537#issuecomment-262865768) discovered a few abandoned crates that weren't updated since `private_in_public` has been introduced in 2015.

cc https://github.com/rust-lang/rust/issues/34537 https://internals.rust-lang.org/t/lang-team-minutes-private-in-public-rules/4504
Fixes https://github.com/rust-lang/rust/issues/30476
Fixes https://github.com/rust-lang/rust/issues/33479

cc @nikomatsakis 
r? @eddyb

petrochenkov commented on src/librustc_privacy/lib.rs:
```diff
@@ -504,6 +504,291 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+                None
+            }
+            ty::TyAnon(..) => {
+                // FIXME: How to visit `impl Trait` type?
```
This needs to be fixed.
IIRC, my attempts to visit it with `TypeVisitor` either infinitely recursed or ICEd.

eddyb replied:
Shouldn't `impl Trait` be able to "export" private types? cc @rust-lang/lang

petrochenkov replied:
Yes, in anonymized form, but I'm talking about preventing use of
```rust
impl PrivateTrait
     ^^^^^^^^^^^^
```
and
```
impl Trait<PrivateType>
           ^^^^^^^^^^^
```
here, not use of private underlying types.

eddyb replied:
Oh. So using API that involves private types? The private types would show up elsewhere if they were involved. Like, using any method on `Trait` would end up with that part in its `Substs`.

petrochenkov commented on src/librustc_privacy/lib.rs:
```diff
@@ -504,6 +504,291 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+                return true;
+            } else if let ty::TyFnDef(..) = ty.sty {
+                // Inherent static methods don't have self type in substs,
+                // we have to check it additionally.
```
For some reason `Self` for static inherent methods is not listed in substs.
IIRC, it's not the first time I see it requiring special treatment for this reason.

eddyb replied:
It's not because `Self` is a mere alias in them. The substitutions for impls (inherent *or* trait impls, but you never get the latter from type-checking, only e.g. trans or MIR inlining can see them) are *literally* for the `<...>` in `impl<...>`.

*However*, it should be impossible to access such an inherent method without one of:
* method call, which means there is an expression containing the `Self` type
* `T::method` which is sugar for `<T>::method`, *and* the semantic version of `T` *can be read* (see `rustdoc` and `rustc_save_analysis`)

petrochenkov commented on src/test/compile-fail/private-inferred-type.rs:
```diff
@@ -0,0 +1,74 @@
+    // ext::Z; // ERROR type `ext::Z` is private
+            // ERROR type `fn(u8) -> ext::Z {ext::Z::{{constructor}}}` is private
+    // ext::Z1; // ERROR type `fn(u8) -> ext::Z1 {ext::Z1::{{constructor}}}` is private
+    // ext::Z1::k; // ERROR type `fn() {ext::Z1::k}` is private
```
Some of these cases cannot be tested until macros 2.0 are merged.

nikomatsakis commented:
@petrochenkov I'm very excited to see this PR! I have a few questions. I think there are some key questions that we were wrestling with:

- Is it legal to define an associated type in a "public impl" that references a private type?
- Is it legal to make a `pub type Foo = PrivateType`? (Presumably it would be illegal to reference it.)
- Is it legal to make a `pub use self::foo::PrivateThing`? (Presumably it would be illegal to reference it.)
- Are these the right questions? Other corner cases or interesting cases I've forgotten about?

I'm sure I'll know the answers once I skim the code, but I thought it'd be helpful to try and summarize these key points so that people don't have to do so.

nikomatsakis commented:
Also, I think we should FCP this before we land it. I don't anticipate a lot of back-and-forth but this is basically an insta-stable language change, so I think it makes sense.

petrochenkov commented:
@nikomatsakis 
>Is it legal to define an associated type in a "public impl" that references a private type?
>Is it legal to make a pub type Foo = PrivateType? (Presumably it would be illegal to reference it.)

I didn't relax the private-in-public checker yet, so these two are still errors. I'm going to send a separate PR with private-in-public changes later. *That* PR will certainly require an FCP, this one - I'm not so sure.

My preliminary plan is that `pub type Foo = PrivateType` will be legal, prohibiting it achieves nothing in presence of macros 2.0 (`pub macro Foo { PrivateType }` does the same thing).
`pub associated_type Foo = PrivateType` is a thorny case, discussed at length in [the internals thread](https://internals.rust-lang.org/t/lang-team-minutes-private-in-public-rules/4504). I think I'll follow the decision from that discussion and keep it an error. It's viable because this check is based on semantic types and traits and not affected by macros. But this is a restriction that can certainly be relaxed in the future if all types are normalized before checking for privacy.

>Is it legal to make a `pub use self::foo::PrivateThing`? (Presumably it would be illegal to reference it.)

This is a separate case unrelated to this PR and its successors, and it will continue being an error.
With [current resolution model](https://github.com/rust-lang/rfcs/blob/master/text/1560-name-resolution.md) this is basically an "unused import"/"`use` importing nothing" error in disguise.
`pub([visibility]) [pattern]` filters names matching `[pattern]` using `[visibility]` and imports what remains. This works for everything - globs, names matching in several namespaces and single names as well. `pub use self::foo::PrivateThing` just filters away all matching names and imports nothing, then it reports an error instead of being useless.

brson commented:
Started builds for crater

- crater https://github.com/rust-lang/rust/pull/42125
  - http://github.com/petrochenkov/rust
  - before 0bd9e1f5e6e9832691d033f1cc32409f5e2a9145
  - https://tools.taskcluster.net/task-inspector/#ucQGJfMgSsav2NSo8LUmXA
  - after 18d9739c5b897cd6a111402ed585d2bfa30601e0
  - https://tools.taskcluster.net/task-inspector/#jfw59nukSfmVTPIIG6oTSw

brson commented:
I've started the crate builds for crater.

brson commented:
https://gist.github.com/anonymous/557f11cdeae48a28a39574a729c74be1

nikomatsakis commented:
Inline (not yet analyzed)

## Coverage

* 9476 crates tested: 7024 working / 2169 broken / 61 regressed / 27 fixed / 195 unknown.

## Regressions

* There are 37 root regressions
* There are 61 regressions

## Root regressions, sorted by rank:

* [rmp-serde-0.13.1](https://crates.io/crates/rmp-serde) ([before](https://tools.taskcluster.net/task-inspector/#RikXfQNUSmO9lfrPt3AcAw)) ([after](https://tools.taskcluster.net/task-inspector/#S6KDsUutTaCFHnT6Tz4yLw))
* [mysql-11.1.0](https://crates.io/crates/mysql) ([before](https://tools.taskcluster.net/task-inspector/#Zbd62CQpTfiIqzS2yFfOEw)) ([after](https://tools.taskcluster.net/task-inspector/#W5D8KwcVRW-IrsfRXJExxw))
* [slog-async-2.0.1](https://crates.io/crates/slog-async) ([before](https://tools.taskcluster.net/task-inspector/#CaOmdLFURT62dLt97zt4tw)) ([after](https://tools.taskcluster.net/task-inspector/#ZRd28qdfS8aZK8HYSJKp-g))
* [mach-0.1.0](https://crates.io/crates/mach) ([before](https://tools.taskcluster.net/task-inspector/#fsYn67eCQL-OMsLPMDm8Mg)) ([after](https://tools.taskcluster.net/task-inspector/#xRWx-_O6SCySACw4Ur6rdw))
* [locale-0.2.1](https://crates.io/crates/locale) ([before](https://tools.taskcluster.net/task-inspector/#BOBT4scsR5exNN0tdydvgg)) ([after](https://tools.taskcluster.net/task-inspector/#8KDPSZpdR7ikrok7M3J98A))
* [measurements-0.4.0](https://crates.io/crates/measurements) ([before](https://tools.taskcluster.net/task-inspector/#VhlN8RM0TEeo04qLtMbRwA)) ([after](https://tools.taskcluster.net/task-inspector/#k7VfhHsjQCqMrZDVKZQiRg))
* [microbench-0.3.2](https://crates.io/crates/microbench) ([before](https://tools.taskcluster.net/task-inspector/#af0khNt3Rt-w229YKvt7-A)) ([after](https://tools.taskcluster.net/task-inspector/#K24ufdXUQ02I2cv1zOIyMg))
* [mush-0.1.2](https://crates.io/crates/mush) ([before](https://tools.taskcluster.net/task-inspector/#qLpNJndwQaCxwiYUfqbVgg)) ([after](https://tools.taskcluster.net/task-inspector/#WIinHGO8Qv6jbuD-d2vwrg))
* [netmap_sys-0.1.3](https://crates.io/crates/netmap_sys) ([before](https://tools.taskcluster.net/task-inspector/#GUnHGcwDRwavOZQ6Kewzsw)) ([after](https://tools.taskcluster.net/task-inspector/#1Qa42kB3TSyLpdFNvXLzMA))
* [google-deploymentmanager2-1.0.4+20161209](https://crates.io/crates/google-deploymentmanager2) ([before](https://tools.taskcluster.net/task-inspector/#wveOvUSCSA6W-gAXlEE1uw)) ([after](https://tools.taskcluster.net/task-inspector/#unruqNosST6-5H8xFuNq6w))
* [secret-service-0.3.1](https://crates.io/crates/secret-service) ([before](https://tools.taskcluster.net/task-inspector/#D2UklxajQh-uK9DdIpc8wg)) ([after](https://tools.taskcluster.net/task-inspector/#qLkz3DtZSzyZFUccAV3S7g))
* [staticdir-0.3.1](https://crates.io/crates/staticdir) ([before](https://tools.taskcluster.net/task-inspector/#kh2PxRAHSly76IY6SVjHVQ)) ([after](https://tools.taskcluster.net/task-inspector/#qtvVi4tPQ8yhBlemGZpoSA))
* [vk_api-1.0.36](https://crates.io/crates/vk_api) ([before](https://tools.taskcluster.net/task-inspector/#PuIbqNCZQOycpKUshjZKHQ)) ([after](https://tools.taskcluster.net/task-inspector/#0av1NPw9RACLWYpfwbsGAQ))
* [hcap-0.0.2](https://crates.io/crates/hcap) ([before](https://tools.taskcluster.net/task-inspector/#oyh_bGARSYKJ1bIH16y2iw)) ([after](https://tools.taskcluster.net/task-inspector/#rA7hyPouQT2dRgLMpo1ZyA))
* [couchdb-0.5.1](https://crates.io/crates/couchdb) ([before](https://tools.taskcluster.net/task-inspector/#2nEi5WfuTgCDXDqOBRatTA)) ([after](https://tools.taskcluster.net/task-inspector/#yEnRwSl_TgeA2MmhZUjSXw))
* [coyoneda-0.5.2](https://crates.io/crates/coyoneda) ([before](https://tools.taskcluster.net/task-inspector/#q777e9vdQ7OGmCD6eFtXpg)) ([after](https://tools.taskcluster.net/task-inspector/#dGFHfKh_Td2AosEzNxuKtA))
* [mango_smoothie-0.2.0](https://crates.io/crates/mango_smoothie) ([before](https://tools.taskcluster.net/task-inspector/#P5wkX3QVS2iL_JSg2f_dcg)) ([after](https://tools.taskcluster.net/task-inspector/#DxBhvdOZTMmS6jz_8DStmQ))
* [math-text-transform-0.1.1](https://crates.io/crates/math-text-transform) ([before](https://tools.taskcluster.net/task-inspector/#J4XNbKa4RwSDajydKMXgPg)) ([after](https://tools.taskcluster.net/task-inspector/#WtWdE4TxQ6e29FMZk5YUPQ))
* [matrixstack-0.1.4](https://crates.io/crates/matrixstack) ([before](https://tools.taskcluster.net/task-inspector/#RaHWNl1RRRKRyEzTePRYxA)) ([after](https://tools.taskcluster.net/task-inspector/#-Yn2dcdmRd-Z6TYMcjwxnA))
* [maybe-owned-0.2.0](https://crates.io/crates/maybe-owned) ([before](https://tools.taskcluster.net/task-inspector/#xwVwnNehQZGPbTyyt12-tQ)) ([after](https://tools.taskcluster.net/task-inspector/#uedeJWWGR0KBvS8RAT34Jg))
* [metagener-0.1.0](https://crates.io/crates/metagener) ([before](https://tools.taskcluster.net/task-inspector/#vGuGuhP2T5i_f73v4-CFBg)) ([after](https://tools.taskcluster.net/task-inspector/#yWwiRv-sRRyuGPkysa8Zkw))
* [mpd-0.0.11](https://crates.io/crates/mpd) ([before](https://tools.taskcluster.net/task-inspector/#KL8AGPTpQrOr07zmIo4nsQ)) ([after](https://tools.taskcluster.net/task-inspector/#OBk4eJmAQdSNmw01w_ahpw))
* [ndarray-parallel-0.3.0](https://crates.io/crates/ndarray-parallel) ([before](https://tools.taskcluster.net/task-inspector/#NQLbUFgZTM6J6K-va-kITQ)) ([after](https://tools.taskcluster.net/task-inspector/#gBnML8FnQ9KDET45BLeN0g))
* [nickel_session-0.1.0](https://crates.io/crates/nickel_session) ([before](https://tools.taskcluster.net/task-inspector/#wGFaAmfuSaKC_HljwzHltA)) ([after](https://tools.taskcluster.net/task-inspector/#GU8MUiIzQx2e0pKR2WHOSA))
* [noise-rust-crypto-0.0.3](https://crates.io/crates/noise-rust-crypto) ([before](https://tools.taskcluster.net/task-inspector/#Vf-3nQ-ARniZ79pZigdmtA)) ([after](https://tools.taskcluster.net/task-inspector/#lpAtWDMIRjObphWtdGmLdA))
* [nom-operator-0.0.1](https://crates.io/crates/nom-operator) ([before](https://tools.taskcluster.net/task-inspector/#tmFbxiL9Rt2ssFdiE3DtPQ)) ([after](https://tools.taskcluster.net/task-inspector/#x4tyAms1R_6QGKC-p9lCFw))
* [pid_control-0.7.2](https://crates.io/crates/pid_control) ([before](https://tools.taskcluster.net/task-inspector/#_UZBagSNQBeK6g-Bs-9UBw)) ([after](https://tools.taskcluster.net/task-inspector/#-64c7NXVRQOtUVZherDJDQ))
* [reql-0.0.8](https://crates.io/crates/reql) ([before](https://tools.taskcluster.net/task-inspector/#Fn_fz2VzTFykOW3vuRNGNQ)) ([after](https://tools.taskcluster.net/task-inspector/#BcW7NQ7KTs-1VFlF2GD6Xw))
* [rgmk-0.2.0](https://crates.io/crates/rgmk) ([before](https://tools.taskcluster.net/task-inspector/#n6MrmdfVQUCv8CCi0rg_dA)) ([after](https://tools.taskcluster.net/task-inspector/#36sfJerQTl-82XK5xYjdYQ))
* [rules-0.0.2](https://crates.io/crates/rules) ([before](https://tools.taskcluster.net/task-inspector/#lrcNZ2mUSg6SoZnj2Fyoqw)) ([after](https://tools.taskcluster.net/task-inspector/#56BSN7MhRRae3ODo0p6_qA))
* [rustpather-0.1.0](https://crates.io/crates/rustpather) ([before](https://tools.taskcluster.net/task-inspector/#YpknTDKtS8uUHlxfstANYQ)) ([after](https://tools.taskcluster.net/task-inspector/#7t7GpkpAQtaC2gPhYIx5Og))
* [slog-perf-0.1.0](https://crates.io/crates/slog-perf) ([before](https://tools.taskcluster.net/task-inspector/#ivYzoDoyT9an6FoogApG8A)) ([after](https://tools.taskcluster.net/task-inspector/#_TyBoX43S-CmF0HhLDhDxw))
* [stacktrace-0.2.1](https://crates.io/crates/stacktrace) ([before](https://tools.taskcluster.net/task-inspector/#Ieq6w_pTTfW6ja-wmkf70w)) ([after](https://tools.taskcluster.net/task-inspector/#gPOso0w1TYyffivY7DG1KQ))
* [swindon-0.4.0](https://crates.io/crates/swindon) ([before](https://tools.taskcluster.net/task-inspector/#O4sBAvkcTvG_rFXYxiBv8g)) ([after](https://tools.taskcluster.net/task-inspector/#Sw9F2YZzTE28mlHtHWhKbA))
* [transcriptome_translation-0.1.0](https://crates.io/crates/transcriptome_translation) ([before](https://tools.taskcluster.net/task-inspector/#kjJRrEKoSJ6jfnEia3rgdw)) ([after](https://tools.taskcluster.net/task-inspector/#guqTFLu1Tl-7cl5Bqiiczw))
* [unbound-0.4.0](https://crates.io/crates/unbound) ([before](https://tools.taskcluster.net/task-inspector/#CBqf_p8tQqylFpozKkeh5w)) ([after](https://tools.taskcluster.net/task-inspector/#iuvBldi8S8CfbnWFVmUD1g))
* [carboxyl_time-0.0.2](https://crates.io/crates/carboxyl_time) ([before](https://tools.taskcluster.net/task-inspector/#GvVaDWqXRuCGuQhaeNHCWw)) ([after](https://tools.taskcluster.net/task-inspector/#u8xgrFPuSMyjE1rkwIF3Mw))

Mark-Simulacrum commented:
Actual root regressions:

* [rmp-serde-0.13.1](https://crates.io/crates/rmp-serde) ([before](https://tools.taskcluster.net/task-inspector/#RikXfQNUSmO9lfrPt3AcAw)) ([after](https://tools.taskcluster.net/task-inspector/#S6KDsUutTaCFHnT6Tz4yLw))
* [mysql-11.1.0](https://crates.io/crates/mysql) ([before](https://tools.taskcluster.net/task-inspector/#Zbd62CQpTfiIqzS2yFfOEw)) ([after](https://tools.taskcluster.net/task-inspector/#W5D8KwcVRW-IrsfRXJExxw))
* [slog-async-2.0.1](https://crates.io/crates/slog-async) ([before](https://tools.taskcluster.net/task-inspector/#CaOmdLFURT62dLt97zt4tw)) ([after](https://tools.taskcluster.net/task-inspector/#ZRd28qdfS8aZK8HYSJKp-g))
* [locale-0.2.1](https://crates.io/crates/locale) ([before](https://tools.taskcluster.net/task-inspector/#BOBT4scsR5exNN0tdydvgg)) ([after](https://tools.taskcluster.net/task-inspector/#8KDPSZpdR7ikrok7M3J98A))
* [secret-service-0.3.1](https://crates.io/crates/secret-service) ([before](https://tools.taskcluster.net/task-inspector/#D2UklxajQh-uK9DdIpc8wg)) ([after](https://tools.taskcluster.net/task-inspector/#qLkz3DtZSzyZFUccAV3S7g))
* [couchdb-0.5.1](https://crates.io/crates/couchdb) ([before](https://tools.taskcluster.net/task-inspector/#2nEi5WfuTgCDXDqOBRatTA)) ([after](https://tools.taskcluster.net/task-inspector/#yEnRwSl_TgeA2MmhZUjSXw))
* [coyoneda-0.5.2](https://crates.io/crates/coyoneda) ([before](https://tools.taskcluster.net/task-inspector/#q777e9vdQ7OGmCD6eFtXpg)) ([after](https://tools.taskcluster.net/task-inspector/#dGFHfKh_Td2AosEzNxuKtA))
* [mpd-0.0.11](https://crates.io/crates/mpd) ([before](https://tools.taskcluster.net/task-inspector/#KL8AGPTpQrOr07zmIo4nsQ)) ([after](https://tools.taskcluster.net/task-inspector/#OBk4eJmAQdSNmw01w_ahpw))
* [reql-0.0.8](https://crates.io/crates/reql) ([before](https://tools.taskcluster.net/task-inspector/#Fn_fz2VzTFykOW3vuRNGNQ)) ([after](https://tools.taskcluster.net/task-inspector/#BcW7NQ7KTs-1VFlF2GD6Xw))
* [rgmk-0.2.0](https://crates.io/crates/rgmk) ([before](https://tools.taskcluster.net/task-inspector/#n6MrmdfVQUCv8CCi0rg_dA)) ([after](https://tools.taskcluster.net/task-inspector/#36sfJerQTl-82XK5xYjdYQ))
* [rules-0.0.2](https://crates.io/crates/rules) ([before](https://tools.taskcluster.net/task-inspector/#lrcNZ2mUSg6SoZnj2Fyoqw)) ([after](https://tools.taskcluster.net/task-inspector/#56BSN7MhRRae3ODo0p6_qA))
* [rustpather-0.1.0](https://crates.io/crates/rustpather) ([before](https://tools.taskcluster.net/task-inspector/#YpknTDKtS8uUHlxfstANYQ)) ([after](https://tools.taskcluster.net/task-inspector/#7t7GpkpAQtaC2gPhYIx5Og))
* [slog-perf-0.1.0](https://crates.io/crates/slog-perf) ([before](https://tools.taskcluster.net/task-inspector/#ivYzoDoyT9an6FoogApG8A)) ([after](https://tools.taskcluster.net/task-inspector/#_TyBoX43S-CmF0HhLDhDxw))
* [swindon-0.4.0](https://crates.io/crates/swindon) ([before](https://tools.taskcluster.net/task-inspector/#O4sBAvkcTvG_rFXYxiBv8g)) ([after](https://tools.taskcluster.net/task-inspector/#Sw9F2YZzTE28mlHtHWhKbA))
* [carboxyl_time-0.0.2](https://crates.io/crates/carboxyl_time) ([before](https://tools.taskcluster.net/task-inspector/#GvVaDWqXRuCGuQhaeNHCWw)) ([after](https://tools.taskcluster.net/task-inspector/#u8xgrFPuSMyjE1rkwIF3Mw)) -- actual origin in carboxyl

petrochenkov commented:
Analysis:
rmp-serde-0.13.1 - type inference is too smart (#33479)
coyoneda-0.5.2 - actually morphism-0.4.0, type inference is too smart (#33479)
mysql-11.1.0 - type inference is too smart (similar to #33479, but with a trait impl)
locale-0.2.1 - unfixed `private_in_public` in the same crate, fixed on master but unpublished
couchdb-0.5.1 - unfixed `private_in_public` in the same crate, fixed on master but unpublished
mpd-0.0.11 - unfixed `private_in_public` in the same crate, fixed on master but unpublished
rgmk-0.2.0 - unfixed `private_in_public` in the same crate, fixed on master but unpublished
rules-0.0.2 - unfixed `private_in_public` in the same crate, not fixed on master
rustpather-0.1.0 - unfixed `private_in_public` in the same crate, the repository doesn't exist anymore
swindon-0.4.0 - allow(`private_in_public`) in the same crate, not fixed on master
slog-async-2.0.1 - allow(`private_in_public`) in the same project (but other crate, slog)
slog-perf-0.1.0 - allow(`private_in_public`) in the same project (but other crate, slog)
reql-0.0.8 - allow(`private_in_public`) in another project (slog)
secret-service-0.3.1 - actually rust-gmp-0.2.10 (old version), unfixed `private_in_public` in the same crate
carboxyl_time-0.0.2 - actually carboxyl-0.1.2 (old version), unfixed `private_in_public` in the same crate

I think it'll be okay to send PRs with fixes and/or version bumps, and then land this without additional warnings.

nikomatsakis commented:
> unfixed private_in_public in the same crate, fixed on master but unpublished

what does "unfixed" mean? oh, unfixed in *that* crate, you mean? (i.e., not on rustc master)

petrochenkov commented:
@nikomatsakis 
>unfixed in that crate, you mean?

Yes

petrochenkov commented:
PRs/issues are submitted to all affected crates.

arielb1 commented:
>  this significantly restricts what macros 2.0 (as implemented in #40847) can do (sorry @jseyfried) - they still can use private names, but can't use private types.

It's hard to offer a strong boundary using private types in macros anyway, because of inference leakages. Unreachable-but-public types can be used if you want to hide a type name.

petrochenkov commented:
Updated with a fix for `impl Trait`, somewhat de-duplicated diagnostics and more tests

eddyb commented on src/librustc_privacy/lib.rs:
```diff
@@ -535,6 +535,324 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+        self.def_id_visibility(did).is_accessible_from(self.current_item, self.tcx)
+    }
+
+    fn check_expr_pat_type(&mut self, id: ast::NodeId, span: Span) -> bool {
```
Shouldn't this also check `node_substs`?

petrochenkov replied:
`node_types` seem to always contain "substituted" types, so it's not necessary to check substitutions again.
(I assume you meant `item_substs` from tables, because `node_substs` don't exist anywhere.)

eddyb replied:
Ah, I was thinking of what it ended up being on a branch. Yes, `item_substs`.
If you do `<Private as PubTrait>::CONST`, `item_substs` will be the *only* place where `Private` is recorded. With functions, it *happens* that the substitutions go into the function type, which is maybe what you're thinking of?

petrochenkov replied:
I see, thanks for the example, I'll add it to tests.
Maybe checking `item_substs` will help to eliminate special treatment of inherent static methods as well? Need to check.

eddyb replied:
Methods have `TyFnDef` types, so I'm afraid you won't see any difference then. But did you see https://github.com/rust-lang/rust/pull/42125#discussion_r117616152? (there's also https://github.com/rust-lang/rust/pull/42125#discussion_r118276538 about `impl Trait` but I'm okay with the current solution for that)

petrochenkov replied:
Regarding https://github.com/rust-lang/rust/pull/42125#discussion_r118276538 I would prefer uniform rules - an object of private type causes immediate error, not delayed error when you call some method on it (maybe you don't).
Objects of private types can be used in other undesirable ways in addition to calling their methods, for example passed to something checking for predicates like `PrivTy: Trait`. This is bad because `PrivTy` should be able to un-implement `Trait` freely at any moment.

petrochenkov replied:
I read https://github.com/rust-lang/rust/pull/42125#discussion_r117616152 earlier, but I still had hope :)
I implemented something working to extract `T` from `<T>::method` but I didn't look at what `rustdoc` and `rustc_save_analysis` do.

petrochenkov replied:
FIXME: Extend the special logic for static inherent methods to inherent associated constants.

petrochenkov replied:
>`T::method` which is sugar for `<T>::method`, and the semantic version of `T` can be read (see `rustdoc` and `rustc_save_analysis`)

I can't find anything in `rustdoc` and `rustc_save_analysis` doing this.

eddyb replied:
They do it for `TyPath` for an associated type `QPath` I think?

eddyb replied:
Will link later when I get home if you don't stumble across it by then.

eddyb replied:
Ah drat, I forgot @nikomatsakis killed it ([what they do now](https://github.com/rust-lang/rust/blob/4ed2edaafe82fb8d44e81e00ca3e4f7659855ba2/src/librustdoc/clean/mod.rs#L1783) is not actually useful).

eddyb replied:
>  an object of private type causes immediate error

IMO, `impl PublicTrait + PrivateTrait` should be visible as `impl PublicTrait` from the outside, and the privacy of the `impl Trait` type *itself* should be the privacy of the defining item (`fn` which returns it, right now).

> for example passed to something checking for predicates like `PrivTy: Trait`

That's a very good reason to keep the conservative scheme you got going, I agree.
I hope we can maybe do something more clever in the future, but there's a limit to that.

eddyb commented on src/librustc_privacy/lib.rs:
```diff
@@ -535,6 +535,324 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+                            self.tcx.sess.span_err(self.span, &msg);
+                            return true;
+                        }
+                        // Skip `Self` to avoid infinite recursion
```
Could you `assert_eq!(trait_ref.self_ty(), ty);`? Other than that, this looks good (and we can relax it in the future).

eddyb replied:
On second thought, given the failures, don't bother. just mention that `Self` is the `TyAnon` `ty` explicitly.

petrochenkov commented:
@eddyb 
Updated with assert from https://github.com/rust-lang/rust/pull/42125#discussion_r119533012

petrochenkov commented:
>Could you assert_eq!(trait_ref.self_ty(), ty);?

Ha-ha.
>thread 'rustc' panicked at 'assertion failed: `(left == right)` (left: `impl std::iter::Iterator`, right: `impl std::iter::Iterator`)', /checkout/src/librustc_privacy/lib.rs:831

eddyb commented:
@petrochenkov Oh are they not substituted? Don't forget to substitute (I'll check in a few hours).

eddyb commented:
Oh also don't forget to check adjustments (note that #42281 has landed so you'll have to rebase first).

petrochenkov commented:
@eddyb 
>Oh also don't forget to check adjustments (note that #42281 has landed so you'll have to rebase first).

Yeah, that was in the plans (and I've seen https://github.com/rust-lang/rust/pull/42281), but I planned to do it later (something private in the middle of deref chain is a pretty contrived setup). Maybe I'll include it in this PR, but not sure - it's desirable to minimize time between landing https://github.com/rust-lang/rust/pull/40847 and this PR.

eddyb commented on src/librustc_privacy/lib.rs:
```diff
@@ -535,6 +535,325 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+            }
+            hir::ExprMethodCall(name, ..) => {
+                // Method calls have to be checked specially.
+                if let Some(method) = self.tables.method_map.get(&ty::MethodCall::expr(expr.id)) {
```
Bonus: if you check substitutions, most of this goes away (with #42281). You *might* have to check the `type_dependent_defs` entry as well, the same way you do static methods, not sure if typeck stops those cases from happening.

eddyb commented on src/librustc_privacy/lib.rs:
```diff
@@ -535,6 +535,325 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+                    // Inherent static methods don't have self type in substs,
+                    // we have to check it additionally.
+                    let mut impl_def_id = None;
+                    if let Some(node_id) = self.tcx.hir.as_local_node_id(def_id) {
```
Nice hack, however it'd be cleaner to use `opt_associated_item`.

petrochenkov replied:
Now I'm not sure this check is entirely correct, I use `self_ty = self.tcx.type_of(impl_def_id)`, but there also may be substitutions in addition to it?

eddyb replied:
The substitutions are those of the method, which... you already look at.

petrochenkov commented:
Updated with fixes for the prefiously discussed stuff.

petrochenkov commented:
There are some remaining issues though.

Consider this setup:
```rust
mod m {
    struct Priv;
    pub struct Pub<T>(pub T);

    pub type PubAlias = Priv;
    pub type PubAlias2 = Pub<Priv>;
}
```
Any use of `PubAlias` (and `PubAlias2`) outside of `m` should report an error.
(At least because `PubAlias` can be passed to something checking for `Priv: Trait`, the same arguments as with objects.)

Issue 1:
```rust
m::PubAlias2 { 0: loop {} }
```
For some reason this expression has type `Pub<()>` and not `Pub<Priv>`. This seems to be some
orthogonal issue with diverging types (some kind of fallback to `()`?), but the result is that the error is not reported.

Issue 2:
```rust
let x: <m::PubAlias as Trait>::AssocType;
```
The type of pattern `x` is checked, but it may have no traces of `m::PubAlias` in it.
Theoretically, something like this may be possible as well:
```rust
// This fails to compile with "error[E0091]: type parameter `T` is unused", but maybe there's some other way?
type Discard<T> = u8;
// `m::PubAlias` doesn't end up in the semantic type and is not checked
fn f() -> Discard<m::PubAlias> {}
```
If there's a way to create something like `Discard<T>` resulting `T` being ignored and not kept in the semantic type, then `Discard<m::PubAlias>` can be written in any type position.

It seems to be okay to leave both issues for later.
I think maybe issue 2 can be fixed by moving the privacy checks for explicitly written types (as opposed to expression/pattern types) to astconv and performing them during conversion of syntactic types into semantic types.

For reference, all contexts where types can be used and need to be checked:
1. types in item interfaces (only "final" semantic types are checked)
2. type annotation on `let` (checked by pattern checks)
3. `as` and type ascription expressions (checked by expr checks)
4. generic arguments in paths and method calls, including associated types.
5. `TYPE` in `<TYPE>::assoc` and `<TYPE as Trait>::assoc`, including associated types.
6. nested syntactic types inside other types in all contexts (not checked, see 4 and 5)

eddyb commented:
Part of the problem with checking the types resulting from syntactic ones is inference.
Ideally *all* the syntactic type nodes would have associated entries in `TypeckTables` that one can just look up, and outside of a body you can arguably use `hir_ty_to_ty`. cc @nikomatsakis

petrochenkov commented:
@eddyb 
>Ideally all the syntactic type nodes would have associated entries in TypeckTables that one can just look up

That's what I thought as well, this would be better *in general*, but all the recent "incremental/on demand" infrastructure changes seem to move things in the opposite direction, so I thought maybe there are some deep reasons for not doing this (the general direction of the compiler infrastructure is a complete mystery for me).

eddyb commented:
*Within* a body there is no issue with bringing back that map, it'd just be in `TypeckTables` and not global. Frankly, I think RLS wants it.

Mark-Simulacrum commented:
This is still tagged as needs-crater but I think that happened? Does it need another run?

eddyb commented:
@Mark-Simulacrum I think it'd be useful to do a cargobomb on this (if crater doesn't get fixed), since it got refined a bit and it might find more cases in the wild. Other than that, r=me.

I think we'll fix the limitations here in the future instead of delaying the PR further.
Unless @petrochenkov disagrees, in which case I will try to help with speeding that up.

petrochenkov commented:
>I think we'll fix the limitations here in the future instead of delaying the PR further.

\+

Mark-Simulacrum commented:
@bors try

Getting artifacts for cargobomb

bors commented:
:hourglass: Trying commit 5645122 with merge ab4661d...

frewsxcv commented:
looks like appveyor timed out

https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3350/job/clxdlnk76bo48mie

@bors retry

tomprince commented:
I've  started a cargobomb run on https://github.com/rust-lang/rust/commit/ab4661d5db2fc9060f1fd9703f9442f687291e74

tomprince commented:
The cargobomb results are at [here](https://cargobomb-reports.s3.amazonaws.com/pr-42125/index.html)

petrochenkov commented:
Analysis of "build-fail" regressions:

**Known regressions:**
carboxyl_time-0.0.2 - fixed, not yet published, issue submitted
coyoneda-0.5.2 - morphism v0.4.1 needs fixing, PR submitted
mpd-0.0.11 - fixed, not yet published, issue submitted
rgmk-0.2.0 - fixed, not published, issue submitted, the author rejected to publish
rules-0.0.2 - needs fixing, PR submitted
rustpather-0.1.0 - the repo doesn't exist anymore\
rust-gmp-0.2.10 - fixed, cargobomb tested an out-of-date crate version for some reason
secret-service-0.3.1 - fixed, cargobomb tested an out-of-date crate version for some reason


**Non-root regressions with known root regressions:**
bdrck_config-0.1.3 - old rmp-serde v0.12.4
cocaine-0.2.0 - old rmp-serde v0.13.3
dbench-0.1.0 - old mysql v8.0.0
dbmigrate-0.3.3 - old mysql v10.0.0
dbmigrate-lib-0.1.1 - old mysql v10.0.0
keyring-0.1.1 - old rust-gmp v0.2.10
mysql-repo-0.1.0 - old mysql v10.0.0
r2d2_mysql-8.0.0 - old mysql v8.0.0
rabble-0.4.0 - old slog v1.5.2

**Previously uknown regressions:**
elastic-0.12.2 - elastic_types v0.17.1 needs fixing

I haven't looked through the "pinned commit" stuff like `mgattozzi.nightly-rust.35b0a2d6d2adb7d60291e41833f2ab3e0c33439e`

eddyb commented:
@nikomatsakis Can you take a quick look and make sure I haven't missed any potential blockers?
Otherwise, r=me.

nikomatsakis commented:
Will do. Sorry for lag.

alexcrichton commented:
@bors: r-

Bors thinks it's testing this?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 5645122 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 56451228a799122a60ca41859993e3de4309ced9 -->

alexcrichton commented:
@bors: r-

bors commented:
:hourglass: Testing commit 56451228a799122a60ca41859993e3de4309ced9 with merge 3639099059902b8526f59bbc1143fd7811d29e03...

alexcrichton commented:
@bors: retry

(sorry trying to get bors under control...)

eddyb commented:
It got confused because of the try build some time back.

bors commented:
:umbrella: The latest upstream changes (presumably #42495) made this pull request unmergeable. Please resolve the merge conflicts.

arielb1 commented:
Hi @petrochenkov what's the status of this PR? Are we waiting for a review by @nikomatsakis? Just trying to make sure it's not getting lost.

petrochenkov commented:
@arielb1 
Yes, waiting for @nikomatsakis

petrochenkov commented:
Rebased.

Ixrec commented on src/test/compile-fail/auxiliary/private-inferred-type.rs:
```diff
@@ -0,0 +1,44 @@
+    pub fn static_method() {}
+}
+impl Pub<u8> {
+    fn priv_metod(&self) {}
```
typo: metod

nikomatsakis commented on src/test/compile-fail/hygiene/fields.rs:
```diff
@@ -22,11 +22,11 @@ mod foo {
 
-        let s = S { x: 0 };
-        let _ = s.x;
+        let s = S { x: 0 }; //~ ERROR type `foo::S` is private
```
What's up with these errors?

petrochenkov replied:
Macro `m` leaks a value of private type `S`, e.g.  the expanded code contains value of private type `S`.

This specific case shouldn't necessarily be an error, because `S` is used "internally" in the macro, and the macro doesn't "return" it to the code outside of it. Discerning between macros using something internally or not is a separate problem that may be solved in the future (or [may be not](https://github.com/rust-lang/rust/pull/42125#issuecomment-304475760)), so this PR implements the conservative solution - everything used in a macro is considered leaked (this is mentioned in the PR description).

nikomatsakis replied:
ok, I see it now -- the macro is used from outside the module, you are saying.

nikomatsakis commented:
( I take it back, GH was showing it to me, at the end of the list, but I didn't notice somehow. :( )

petrochenkov commented:
Rebased, comments addressed.

aidanhs commented:
Pinged @nikomatsakis for review on IRC.

petrochenkov commented:
@nikomatsakis @eddyb @nrc
Macros 2.0 are going ahead with full speed (https://github.com/rust-lang/rust/pull/40847, https://github.com/rust-lang/rust/pull/40939) and this is a prerequisite. It may be not immediately obvious because this prohibits things rather than enables them, but it's still important, to avoid breaking people's code in particular.
One release was already missed since https://github.com/rust-lang/rust/pull/40847 landed and the next release is soon enough, please prioritize landing this PR.

nikomatsakis commented:
@petrochenkov sorry, was on another vacation this week. I sort of thought I had r+'d, but re-reading my comment I see it was ambiguous! I will go ahead and re-read your latest commits now though.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 4fb032e has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 4fb032e6d988baf3c8ee756f38205b799fd1eccd -->

bors commented:
:lock: Merge conflict

petrochenkov commented:
Rebased.
@bors r=nikomatsakis

bors commented:
:pushpin: Commit a27f8cf has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis a27f8cf8a36f3d69267ed25602504144fab1d896 -->

bors commented:
:hourglass: Testing commit a27f8cf8a36f3d69267ed25602504144fab1d896 with merge b80b659d674fa19b2b48142e4f7974544bc8d5f6...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3828), [status-travis](https://travis-ci.org/rust-lang/rust/builds/250997011?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing b80b659d674fa19b2b48142e4f7974544bc8d5f6 to master...

alexcrichton commented:
@rust-lang/compiler just FYI but there's 5 open regressions against beta for this PR, all linked above this comment.
